File: Python/stacks_and_queues/fixed_multi_stack.cpp
#include <iostream>
#include <cstdlib>
#include <cassert>

class FixedMultiStack
{
public:
  //Create 3 stacks, each stack is of size stackCapacity.
  FixedMultiStack(int stackCapacity);
  virtual ~FixedMultiStack();

  //Push an element onto stack stackNum, where stackNum is from 0 to 2.
  void push(int stackNum, int value);

  //Pop the top element from stack stackNum, where stackNum is from 0 to 2.
  void pop(int stackNum);

  //Return the top element on stack stackNum, where stackNum is from 0 to 2.
  int top(int stackNum) const;

  bool isEmpty(int stackNum) const;
  bool isFull(int stackNum) const;

private:
  int numOfStack = 3;
  int stackCapacity;
  int *stackArray;
  int *stackCapacityUsed;

  //Return the top index of stack stackNum, where stackNum is from 0 to 2.
  int indexOfTop(int stackNum) const;
};


FixedMultiStack::FixedMultiStack(int stackCapacity)
{
  this->stackCapacity = stackCapacity;
  stackArray = new int[numOfStack * stackCapacity]();
  stackCapacityUsed = new int[numOfStack]();
}

FixedMultiStack::~FixedMultiStack()
{
  delete [] stackArray;
  delete [] stackCapacityUsed;
}

void FixedMultiStack::push(int stackNum, int value)
{
  if(isFull(stackNum))
    std::cout << "Stack " << stackNum << " is full.\n";
  else
  {
    stackCapacityUsed[stackNum]++;
    stackArray[indexOfTop(stackNum)] = value;
  }
}

void FixedMultiStack::pop(int stackNum)
{
  if(isEmpty(stackNum))
    std::cout << "Stack " << stackNum << " is empty.\n";
  else
  {
    int topIndex = indexOfTop(stackNum);
    stackArray[topIndex] = 0;
    stackCapacityUsed[stackNum]--;
  }
}

int FixedMultiStack::top(int stackNum) const
{
  if(isEmpty(stackNum))
  {
    std::cout << "Stack " << stackNum << " is empty.\n";
    exit(1); //Or throw an exception.
  }
  else
    return stackArray[indexOfTop(stackNum)];
}

bool FixedMultiStack::isEmpty(int stackNum) const
{
  return (stackCapacityUsed[stackNum] == 0);
}

bool FixedMultiStack::isFull(int stackNum) const
{
  return (stackCapacityUsed[stackNum] == stackCapacity);
}

int FixedMultiStack::indexOfTop(int stackNum) const
{
  int startIndex = stackNum * stackCapacity;
  int capacity = stackCapacityUsed[stackNum];
  return (startIndex + capacity - 1);
}


int main() {
  const int stackCapacity = 5;
  FixedMultiStack fms{stackCapacity};

  // Test pushing and popping from stack 0
  fms.push(0, 10);
  fms.push(0, 20);
  fms.push(0, 30);
  assert(fms.top(0) == 30);
  fms.pop(0);
  assert(fms.top(0) == 20);
  fms.pop(0);
  assert(fms.top(0) == 10);
  fms.pop(0);
  assert(fms.isEmpty(0));

  // Test pushing and popping from stack 1
  fms.push(1, 40);
  fms.push(1, 50);
  fms.push(1, 60);
  assert(fms.top(1) == 60);
  fms.pop(1);
  assert(fms.top(1) == 50);
  fms.pop(1);
  assert(fms.top(1) == 40);
  fms.pop(1);
  assert(fms.isEmpty(1));

  // Test pushing and popping from stack 2
  fms.push(2, 70);
  fms.push(2, 80);
  fms.push(2, 90);
  assert(fms.top(2) == 90);
  fms.pop(2);
  assert(fms.top(2) == 80);
  fms.pop(2);
  assert(fms.top(2) == 70);
  fms.pop(2);
  assert(fms.isEmpty(2));

  // Test that pushing onto a full stack returns an error message
  for (int i = 0; i < stackCapacity; i++) {
    fms.push(0, i);
  }
  assert(fms.isFull(0));
  std::cout << "Attempting to push onto a full stack...\n";
  fms.push(0, stackCapacity); // This should return an error message

  // Test that popping from an empty stack returns an error message
  assert(fms.isEmpty(0));
  std::cout << "Attempting to pop from an empty stack...\n";
  fms.pop(0); // This should return an error message

  return 0;
}

File: Python/stacks_and_queues/fixed_multi_stack.py
import pytest

class MultiStack:
    """
    A class representing a multi-stack data structure, consisting of multiple
    stacks stored in a single array.
    """
    def __init__(self, stack_size: int, number_of_stacks: int) -> None:
        """
        Initializes a new MultiStack object with the specified parameters.

        :param stack_size: an integer representing the maximum size of each stack
        :param number_of_stacks: an integer representing the total number of stacks
        """
        self.number_of_stacks = number_of_stacks  # store number of stacks
        self.array = [[] for _ in range(number_of_stacks)]  # initialize an empty list for each stack
        self.sizes = [0] * self.number_of_stacks
        self.stack_size = stack_size  # store the maximum size of each stack


    def push(self, value: int, stack_num: int) -> None:
        """
        Pushes the given value onto the specified stack.

        :param value: The value to be pushed onto the stack.
        :param stack_num: The number of the stack onto which the value is to be pushed.
        :raises StackFullError: If the specified stack is already full.
        """
        self._assert_valid_stack_num(stack_num)
        if self.is_full(stack_num):
            raise StackFullError(f"Push failed: stack #{stack_num} is full")
        # Append the value to the appropriate stack in the array
        self.array[stack_num].append(value)
        # Increment the size of the stack
        self.sizes[stack_num] += 1

    def pop(self, stack_num: int) -> int:
        """
        Pops the top element from the specified stack and returns it.
        
        @param stack_num: the stack number to pop from (0, 1, or 2)
        @return the popped value
        @throws StackEmptyError if the specified stack is empty
        """
        self._assert_valid_stack_num(stack_num)  # validate stack_num is within range
        if self.is_empty(stack_num):
            raise StackEmptyError(f"Cannot pop from empty stack #{stack_num}")
        return self.array[stack_num].pop()  # pop the top element from the specified stack and return it


    def peek(self, stack_num: int) -> int:
        """
        Returns the top element of the specified stack without removing it.
        
        @param stack_num: the stack number to peek at (0, 1, or 2)
        @return the top element of the specified stack
        @throws StackEmptyError if the specified stack is empty
        """
        self._assert_valid_stack_num(stack_num)  # validate stack_num is within range
        if self.is_empty(stack_num):
            raise StackEmptyError(f"Cannot peek at empty stack #{stack_num}")
        return self.array[stack_num][-1]  # return the top element of the specified stack

    def is_empty(self, stack_num: int) -> bool:
        """
        Returns True if the specified stack is empty, False otherwise.
        
        @param stack_num: the stack number to check (0, 1, or 2)
        @return True if the specified stack is empty, False otherwise
        """
        self._assert_valid_stack_num(stack_num)  # validate stack_num is within range
        return len(self.array[stack_num]) == 0  # check if the specified stack is empty and return the result


    def is_full(self, stack_num: int) -> bool:
        """
        Returns True if the specified stack is full, False otherwise.
        
        @param stack_num: the stack number to check (0, 1, or 2)
        @return True if the specified stack is full, False otherwise
        """
        self._assert_valid_stack_num(stack_num)  # validate stack_num is within range
        return len(self.array[stack_num]) == self.stack_size  # check if the specified stack is full and return the result

    def _assert_valid_stack_num(self, stack_num: int) -> None:
        """
        Helper method that raises an exception if the given stack number is invalid.
        
        @param stack_num: the stack number to check (0 to number_of_stacks-1)
        @throws StackDoesNotExistError if the specified stack does not exist
        """
        if stack_num >= self.number_of_stacks:  # check if stack_num is out of range
            raise StackDoesNotExistError(f"Stack #{stack_num} does not exist")  # raise an exception if stack_num is invalid


class MultiStackError(Exception):
    """multistack operation error"""


class StackFullError(MultiStackError):
    """the stack is full"""


class StackEmptyError(MultiStackError):
    """the stack is empty"""


class StackDoesNotExistError(ValueError):
    """stack does not exist"""


def test_multistack():
    num_stacks = 3
    stack_size = 6
    s = MultiStack(stack_size=stack_size, number_of_stacks=num_stacks)

    for stack_num in range(num_stacks):
        assert s.is_empty(stack_num)
        assert not s.is_full(stack_num)
        with pytest.raises(StackEmptyError):
            s.pop(stack_num)

        for i in range(stack_size - 1):
            s.push(i, stack_num=stack_num)
            assert s.peek(stack_num) == i
            assert not s.is_empty(stack_num)
            assert not s.is_full(stack_num)

        s.push(999, stack_num=stack_num)
        with pytest.raises(StackFullError):
            s.push(777, stack_num=stack_num)

        assert not s.is_empty(stack_num)
        assert s.is_full(stack_num)

        assert s.peek(stack_num) == 999
        assert s.pop(stack_num) == 999
        assert not s.is_empty(stack_num)
        assert not s.is_full(stack_num)

        for i in range(stack_size - 2, 0, -1):
            assert s.peek(stack_num) == i
            assert s.pop(stack_num) == i
            assert not s.is_empty(stack_num)
            assert not s.is_full(stack_num)

        assert s.peek(stack_num) == 0
        assert s.pop(stack_num) == 0
        assert s.is_empty(stack_num)
        assert not s.is_full(stack_num)

        with pytest.raises(StackEmptyError):
            s.peek(stack_num)
        with pytest.raises(StackEmptyError):
            s.pop(stack_num)


def test_stack_does_not_exist():
    s = MultiStack(stack_size=3, number_of_stacks=1)
    with pytest.raises(StackDoesNotExistError):
        s.push(1, 1)


if __name__ == "__main__":
    newstack = MultiStack(2, 2)
    print(newstack.is_empty(1))
    newstack.push(3, 1)
    print(newstack.peek(1))
    print(newstack.is_empty(1))
    newstack.push(2, 1)
    print(newstack.peek(1))
    print(newstack.pop(1))
    print(newstack.peek(1))
    newstack.push(3, 1)